pkg/controller/bastion/actuator.go

```diff
@@ -0,0 +1,285 @@
+}
+
+func createSecurityGroup(openstackClientFactory openstackclient.Factory, createOpts groups.CreateOpts) (*groups.SecGroup, error) {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return nil, err
+	}
+	return client.CreateSecurityGroup(createOpts)
+}
+
+func deleteSecurityGroup(openstackClientFactory openstackclient.Factory, groupid string) error {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return err
+	}
+	return client.DeleteSecurityGroup(groupid)
+}
+
+func getSecurityGroups(openstackClientFactory openstackclient.Factory, name string) ([]groups.SecGroup, error) {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return nil, err
+	}
+	return client.GetSecurityGroupbyName(name)
+}
+
+func createRules(openstackClientFactory openstackclient.Factory, createOpts rules.CreateOpts) (*rules.SecGroupRule, error) {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return nil, err
+	}
+	return client.CreateRule(createOpts)
+}
+
+func listRules(openstackClientFactory openstackclient.Factory, secGroupID string) ([]rules.SecGroupRule, error) {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return nil, err
+	}
+
+	listOpts := rules.ListOpts{
+		SecGroupID: secGroupID,
+	}
+	return client.ListRules(listOpts)
+}
+
+func deleteRule(openstackClientFactory openstackclient.Factory, ruleID string) error {
+	client, err := openstackClientFactory.Networking()
+	if err != nil {
+		return err
+	}
+	return client.DeleteRule(ruleID)
+}
+
+func listFlavors(openstackClientFactory openstackclient.Factory, listOpts flavors.ListOpts) ([]flavors.Flavor, error) {
+	client, err := openstackClientFactory.Compute()
+	if err != nil {
+		return nil, err
+	}
+	return client.ListFlavors(listOpts)
+}
+
+func setDefaultFlavor(openstackClientFactory openstackclient.Factory) (string, error) {
+	flavors, err := listFlavors(openstackClientFactory, flavors.ListOpts{})
+	if err != nil {
+		return "", err
+	}
+
+	flavorId := ""
+	var min int
+	for _, flavor := range flavors {
+		if flavor.VCPUs >= 2 && isNum(flavor.ID) {
+			if min == 0 || flavor.RAM <= min {
+				min = flavor.RAM
+				flavorId = flavor.ID
+			}
+		}
+	}
+	return flavorId, nil
+}
+
+func isNum(s string) bool {
+	_, err := strconv.ParseInt(s, 10, 0)
+	return err == nil
+}
+
+func listImages(openstackClientFactory openstackclient.Factory, listOpts images.ListOpts) ([]images.Image, error) {
+	client, err := openstackClientFactory.Compute()
+	if err != nil {
+		return nil, err
+	}
+	return client.ListImages(listOpts)
+}
+
+func setDefaultImage(openstackClientFactory openstackclient.Factory) (string, error) {
+	images, err := listImages(openstackClientFactory, images.ListOpts{})
+	if err != nil {
+		return "", err
+	}
+
+	imageId := ""
+	var min int
+	for _, image := range images {
+		if min == 0 || image.MinRAM <= min {
+			if osType, ok := image.Metadata["vmware_ostype"]; ok {
```

Comment:
Is this not again a sap specific assumption?
Why not simply relying on the image config at all?
From my point of view we have two options:
1. We validate the bastion config part of the extension config on extension startup and let it fail if the config does not contain a image/machine type for the bastion
2. We default the config for bastion image/machine type to ensure that there is an empty string and let the bastion operation fail in case there is no config for the bastion image/machine type

Reply:
sorry didn't aware of Metadata at that time. Thanks. I set an empty string in  a110adb3 now when operation reconcile.

Reply:
@dkistner and @petersutter  any comments for the change  https://github.com/gardener/gardener-extension-provider-openstack/commit/1b9d070c6b1dea4d5f4773015f859f403b631a47 Thanks

Reply:
Looks now good to me. The bastion reconcile operation will fail in case that the extension component config does not contain a flavour and/or image reference.
I guess this is good as we let only the `bastion` controller fail in case of missing configuration.
Just a nit: You could rephrase the error message in case the image or flavour is not there:

`s/bastionComputeConfig.imageRef && bastionComputeConfig.flavorRef must not be empty in 00-componentconfig.yaml/bastion not supported as no image and flavour is configured for the bastion host machine/`

Reply:
sorry, the wrong hash code in here I was pasted, it should be https://github.com/gardener/gardener-extension-provider-openstack/pull/365/commits/a56909d614e84bf19adab25468dcdc36c603d8e1. and the error code changed as `"bastionConfig.flavorRef must not be empty"` and `"bastionConfig.ImageRef must not be empty"`.  I will change to `bastion not supported as no image and flavour is configured for the bastion host machine` individually and rebase the PR